### ISSUE TITLE
CRIU Kernel Patches

### ISSUE THREAD

**Issue by sinainnet** [checkpoint-restore/criu#421]
Hi,

I'm curious to know what are exactly the supports Linux Kernel provides for the CRIU. I've already searched the Kernel tree, but I think that I do not know where to search!

Can any one kindly help me?

**Comment by adrianreber** [comment]
There is a list of all CRIU related kernel patches at https://criu.org/Upstream_kernel_commits

**Comment by sinainnet** [comment]
Thank you.

**Comment by sinainnet** [comment]
This listing is great. However, I want to know what exactly happen when fro example, the memory of a process is going to be dumped. I know that the user space part ask a CRIU service to dump it. Then, the service tries for building a protobuf message! After that, it is hard to trace what would happen. I think this part is is implemented in the kernel space, but I could not find it.

**Comment by adrianreber** [comment]
Unfortunately I do not fully understand your question but not much happens in kernel space. Could you rephrase your question?

**Comment by sinainnet** [comment]
Excuse me for bad writing,

Suppose a process contains 10 GBytes of memory. Writing all this memory to a file is time consuming. So, my question is: what exactly happen for checkpointing/restoring allocated memory of a process?

**Comment by adrianreber** [comment]
So, time consuming is of course relative 😉

As this is a bug tracker it does not feel like the right place to discuss CRIU internals and I would recommend going through the documentation provided at https://criu.org/ and also through the source code for more details. There are also some presentations discussing the basic functionality of CRIU.

But the basic concept of CRIU's checkpointing is to suspend the process, insert CRIU's 'parasite code' into the process which dumps the memory from within the to be checkpointed process and write it to disk. If looking at process migration the time to write the memory to the disk is anyway much faster (depending on the disk) than the time required to transfer the memory to the destination system.
